### PR TITLE
internal: add EventTime to timestamp filtering

### PR DESCRIPTION
**Pull request by dgrisonnet**
In the case of a singleton Event generated by `k8s.io/client-go/tools/events` EventRecorder, the Event creation Timestamp is only set inside the EventTime field. Leaving both the Series and the LastTimestamp nil. Thus, we also need to check the EventTime before updating kube_events_total counter.

### CODE REVIEW

File: internal/collector/event.go [hunk 1]
```diff
@@ -100,7 +100,15 @@ func beforeLastEvent(t time.Time, ev *v1.Event) bool {
 		return t.Before(ev.Series.LastObservedTime.Time)
 	}
 
-	return t.Before(ev.LastTimestamp.Time)
+	if !ev.LastTimestamp.IsZero() {
```
Comment (lilic) [inline internal/collector/event.go:103]: I would like you to add a unit test for this function `beforeLastEvent`, you could easily unit test it with the various edge cases it also might help us find a few other edge cases.

Comment (dgrisonnet) [reply internal/collector/event.go]: @lilic I have added some test scenarios in my last commit, would you mind having another look?

File: internal/collector/event.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func (collector *EventCollector) Run(stopCh <-chan struct{}) {
 		AddFunc: func(obj interface{}) {
 			ev := obj.(*v1.Event)
 			// Count only Events created after the exporter starts running.
-			if beforeLastEvent(startRunning, ev) {
+			if beforeLatestEvent(startRunning, ev) {
 				// FIXME: take into account the event count.
```
Comment (lilic) [inline internal/collector/event.go:72]: Is this still a valid TODO?

File: internal/collector/event_test.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestBeforeLatestEvent(t *testing.T) {
+	now := time.Now()
```
Comment (lilic) [inline internal/collector/event_test.go:28]: This should not be time.Now() but a fixed time in the past. e.g. Using `time.Date()` https://golang.org/pkg/time/#Date this makes sure there are no flakes and time is fixed.

File: internal/collector/event.go (hunk 2 above)
Comment (dgrisonnet) [reply internal/collector/event.go]: No, I removed it. Thanks for noticing it :)

**Comment by dgrisonnet** [comment]
@lilic thank you for the review! :slightly_smiling_face:  I addressed your comments, please have another look.

File: internal/collector/event_test.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			Desc: "only newer LastObservedTime",
+			Event: &v1.Event{
+				Series: &v1.EventSeries{LastObservedTime: metav1.NewMicroTime(newer)},
+			},
+			Expected: true,
+		},
+		{
+			Desc: "only older time instants",
+			Event: &v1.Event{
+				FirstTimestamp: metav1.NewTime(older),
```
Comment (lilic) [inline internal/collector/event_test.go:60]: Could we have a case for:
- newer, older, now
- newer, older, older
- etc?

We just want to cover all cases, unless you think these are the only ones that we need to cover?

Comment (dgrisonnet) [reply internal/collector/event_test.go:60]: I don't think it is meaningful to cover all cases tbh. There are a lot of scenarios that we could provide to this function but in practice, there will only be 4-5 scenarios.
For this test, I did not provide real Event scenarios as those will most likely be tested in e2e tests. However, the scenario I came with were: "one newer timestamp", "only one newer timestamp field provided", "no newer timestamp", "only one older timestamp field provided". What do you think?

File: internal/collector/event.go [hunk 5]
```diff
@@ -95,12 +94,25 @@ func (collector *EventCollector) increaseEventsTotal(event *v1.Event, nbNew floa
 	}).Add(nbNew)
 }
 
-func beforeLastEvent(t time.Time, ev *v1.Event) bool {
-	if ev.Series != nil && !ev.Series.LastObservedTime.IsZero() {
-		return t.Before(ev.Series.LastObservedTime.Time)
+func beforeLatestEvent(t time.Time, ev *v1.Event) bool {
+	eventTimes := []time.Time{
+		ev.FirstTimestamp.Time,
```
Comment (lilic) [inline internal/collector/event.go:99]: what if event is nil?

File: internal/collector/event_test.go (hunk 4 above)
Comment (lilic) [reply internal/collector/event_test.go:60]: > function but in practice, there will only be 4-5 scenarios.

I don't see 4-5 scenarios here. It's not urgent to add them, but nice to add have more of them. It's not blocking this PR.

File: internal/collector/event.go (hunk 5 above)
Comment (dgrisonnet) [reply internal/collector/event.go:99]: It should never be nil as the Events are provided by the informer event handler.

**Comment by dgrisonnet** [comment]
@lilic I changed the test cases to have real case scenarios and split the different scenarios into sub-tests that are run in parallel. Please have another look :slightly_smiling_face:
